File: packages/sdk/src/p256.ts
// P-256 ECDSA verification using WebCrypto.
//
// Receipts and attestations carry their signatures DER-encoded
// (matches Apple CryptoKit's wire format on macOS, and what the
// `p256` Rust crate produces on other platforms). WebCrypto's
// SubtleCrypto.verify only accepts raw r||s (64 bytes), so we
// decode the DER ourselves before delegating.
//
// SubtleCrypto handles the SHA-256 pre-hash internally when we ask
// for `{ name: 'ECDSA', hash: 'SHA-256' }`, which matches the Rust
// `p256::ecdsa::Signer for SigningKey` default and Apple's
// `P256.Signing.PrivateKey.signature(for:)`.

import { canonicalize } from "./canonical.ts";

export class SignatureVerifyError extends Error {
  readonly code: string;
  constructor(code: string, message: string) {
    super(message);
    this.code = code;
    this.name = "SignatureVerifyError";
  }
}

/** Verify a P-256 ECDSA signature over `message` with the given
 *  public key. Both inputs use base64 in transit; the message is
 *  raw bytes as produced by [`canonicalize`]. */
export async function verifyP256(
  publicKeyB64: string,
  signatureDerB64: string,
  message: Uint8Array,
): Promise<boolean> {
  const pubRaw = decodeBase64(publicKeyB64);
  if (pubRaw.byteLength !== 64) {
    throw new SignatureVerifyError(
      "bad-pubkey-length",
      `expected 64 raw P-256 bytes (X||Y), got ${pubRaw.byteLength}`,
    );
  }
  const sigDer = decodeBase64(signatureDerB64);
  const sigRaw = derToRawSignature(sigDer);
  const cryptoKey = await crypto.subtle.importKey(
    "raw",
    // Copy into a plain ArrayBuffer-backed view: TS 5.7+ types the generic
    // `Uint8Array<ArrayBufferLike>` as not assignable to WebCrypto's
    // `BufferSource` (which wants `ArrayBuffer`). Runtime-equivalent.
    new Uint8Array(prefixUncompressed(pubRaw)),
    { name: "ECDSA", namedCurve: "P-256" },
    false,
    ["verify"],
  );
  return crypto.subtle.verify(
    { name: "ECDSA", hash: "SHA-256" },
    cryptoKey,
    new Uint8Array(sigRaw),
    new Uint8Array(message),
  );
}

/** Verify a receipt body's `enclaveSignature` against an attestation's
 *  `publicKey`. Strips the signature field AND any `$type` lexicon framing,
 *  canonicalises everything else, and runs WebCrypto. Returns false if the
 *  math fails or any field is missing/malformed.
 *
 *  The `$type` strip is essential and mirrors {@link verifyAttestationSignature}:
 *  the provider signs the receipt body BEFORE the record is written to its PDS,
 *  so the signed canonical bytes never include `$type`. An indexed/stored
 *  record DOES carry `$type` (atproto adds it), and because `$type` sorts to
 *  the front of the canonical JSON, leaving it in shifts every byte and makes
 *  the signature fail. Stripping it is a no-op when absent, so this is always
 *  safe. (Omitting this strip silently rejected every receipt once they began
 *  flowing through the indexer with `$type` populated — the 2026-06 stall.) */
export async function verifyReceiptSignature(
  receipt: { enclaveSignature?: string } & Record<string, unknown>,
  attestationPublicKeyB64: string,
): Promise<boolean> {
  const sig = receipt.enclaveSignature;
  if (!sig) return false;
  const { enclaveSignature: _omit, $type: _type, ...signed } = receipt as Record<string, unknown>;
  const message = new TextEncoder().encode(canonicalize(signed));
  try {
    return await verifyP256(attestationPublicKeyB64, sig, message);
  } catch (e) {
    if (e instanceof SignatureVerifyError) return false;
    throw e;
  }
}

/** Verify an attestation record's `selfSignature` against its own
 *  `publicKey`. This is what authenticates every posture field — `cdHash`,
 *  `getTaskAllow`, `hardenedRuntime`, `encryptionPubKey`, … — as having been
 *  signed by the enclave key, so a verifier MUST run this before trusting any
 *  of them. (The MDA binding only proves `publicKey` is the device's key; the
 *  session-key signature only covers the ephemeral key. Neither covers
 *  posture.) Strips `selfSignature` (and any `$type` lexicon framing),
 *  canonicalises the rest, and runs WebCrypto. */
export async function verifyAttestationSignature(
  attestation: { selfSignature?: string } & Record<string, unknown>,
  publicKeyB64: string,
): Promise<boolean> {
  const sig = attestation.selfSignature;
  if (!sig) return false;
  const { selfSignature: _omit, $type: _type, ...signed } = attestation as Record<string, unknown>;
  // Try the record exactly as stored first (correct for any attestation whose
  // signed canonical bytes equal its stored bytes — the post-fix provider).
  // Fall back to the LEGACY reconstruction for attestations published by the
  // pre-2026-07 provider, whose signed bytes differed from the stored record
  // in two ways (see legacyAttestationBody). Those records can't be re-signed
  // (the signature is enclave-bound), so the verifier reconstructs what was
  // signed. The exact-as-stored form is always tried first, so this fallback
  // never weakens verification of a correctly-signed record — it only rescues
  // records that would otherwise be unverifiable forever.
  for (const body of [signed, legacyAttestationBody(signed)]) {
    const message = new TextEncoder().encode(canonicalize(body));
    try {
      if (await verifyP256(publicKeyB64, sig, message)) return true;
    } catch (e) {
      if (e instanceof SignatureVerifyError) continue;
      throw e;
    }
  }
  return false;
}

/** Reconstruct the canonical body the pre-2026-07 provider actually SIGNED,
 *  given the attestation as STORED on the PDS. Two divergences existed between
 *  what `provider/src/attestation.rs` signed and what it wrote:
 *
 *  1. `mdaCertChain` — the signed `unsigned` JSON always carried
 *     `"mdaCertChain":[]`, but the stored record used
 *     `#[serde(skip_serializing_if = "Vec::is_empty")]`, dropping the field
 *     entirely for the (overwhelmingly common) self-attested case. So restore
 *     `[]` when the field is absent. A NON-empty chain is stored verbatim and
 *     was signed verbatim, so leave it untouched.
 *  2. timestamps — the signed form used seconds-precision RFC3339
 *     (`SecondsFormat::Secs`), but the stored record serialized
 *     `DateTime<Utc>` at chrono's default sub-second precision
 *     (`...:08.384834Z`). Truncate the fractional seconds back off to recover
 *     the signed string.
 *
 *  This mirrors the post-fix provider, which now stores exactly what it signs
 *  (seconds-precision String timestamps, mdaCertChain only when non-empty) —
 *  so post-fix records verify on the as-stored attempt and never reach here. */
function legacyAttestationBody(signed: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = { ...signed };
  if (!("mdaCertChain" in out)) out.mdaCertChain = [];
  for (const key of ["attestedAt", "expiresAt"]) {
    const v = out[key];
    if (typeof v === "string") out[key] = v.replace(/\.\d+(Z|[+-]\d\d:\d\d)$/, "$1");
  }
  return out;
}

// ---- internals -------------------------------------------------------

function decodeBase64(b64: string): Uint8Array {
  // Node 18+ exposes Buffer; we avoid relying on it so this module
  // also works in browsers and edge runtimes.
  const bin = atob(b64);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

function prefixUncompressed(raw64: Uint8Array): Uint8Array {
  const out = new Uint8Array(65);
  out[0] = 0x04;
  out.set(raw64, 1);
  return out;
}

/** Convert a DER-encoded ECDSA signature to the IEEE P-1363 raw
 *  format WebCrypto expects: r (32 bytes) || s (32 bytes), big-endian,
 *  zero-padded. */
export function derToRawSignature(der: Uint8Array): Uint8Array {
  if (der.length < 8 || der[0] !== 0x30) {
    throw new SignatureVerifyError("bad-der", "not a DER SEQUENCE");
  }
  let pos = 2;
  // Length byte may be long-form for very long sigs, but a P-256
  // signature is always short-form.
  if ((der[1]! & 0x80) !== 0) {
    const lenBytes = der[1]! & 0x7f;
    pos = 2 + lenBytes;
  }
  const r = readDerInteger(der, pos);
  const s = readDerInteger(der, r.next);
  const out = new Uint8Array(64);
  out.set(padTo32(r.bytes), 0);
  out.set(padTo32(s.bytes), 32);
  return out;
}

function readDerInteger(der: Uint8Array, pos: number): { bytes: Uint8Array; next: number } {
  if (der[pos] !== 0x02) {
    throw new SignatureVerifyError("bad-der", `expected INTEGER at ${pos}`);
  }
  const len = der[pos + 1]!;
  let start = pos + 2;
  let actualLen = len;
  // Strip a single leading 0x00 padding byte added when the
  // high-order bit of the value would otherwise make it negative.
  if (actualLen > 32 && der[start] === 0x00) {
    start++;
    actualLen--;
  }
  if (actualLen > 32) {
    throw new SignatureVerifyError("bad-der", `INTEGER too long (${actualLen}) for P-256`);
  }
  return { bytes: der.slice(start, start + actualLen), next: start + actualLen };
}

function padTo32(b: Uint8Array): Uint8Array {
  if (b.length === 32) return b;
  if (b.length > 32) {
    throw new SignatureVerifyError("bad-der", "INTEGER > 32 bytes");
  }
  const out = new Uint8Array(32);
  out.set(b, 32 - b.length);
  return out;
}

File: sdk/py/cocore/p256.py
"""P-256 ECDSA verification (mirror of packages/sdk/src/p256.ts).

Signatures are DER-encoded (Apple CryptoKit / the `p256` Rust crate wire format)
and verified with a SHA-256 prehash, over canonical bytes. Public keys are the
raw 64-byte X||Y point, base64 — the encoding the attestation publishes.
"""

from __future__ import annotations

import base64
from typing import Any, Mapping

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.asymmetric.utils import Prehashed, decode_dss_signature

from .canonical import canonical_bytes

# Order n of the P-256 (secp256r1) curve, and n/2 (the low-S ceiling).
_P256_N = 0xFFFFFFFF00000000FFFFFFFFFFFFFFFFBCE6FAADA7179E84F3B9CAC2FC632551
_P256_HALF_N = _P256_N >> 1


def signature_is_high_s(signature_der_b64: str) -> bool:
    """True iff a DER P-256 signature is in the malleable high-S form (s > n/2).

    Parity with the TS ``signatureIsHighS``. Exposed so producers/tests can assert
    their signers normalise to low-S before the ``require_low_s`` guard is enabled
    by default. Returns False for a signature that fails to parse."""
    try:
        _r, s = decode_dss_signature(base64.b64decode(signature_der_b64))
    except (ValueError, TypeError):
        return False
    return s > _P256_HALF_N


def verify_p256(
    public_key_b64: str,
    signature_der_b64: str,
    message: bytes,
    *,
    require_low_s: bool = False,
) -> bool:
    """Verify a DER ECDSA-P256 signature over ``message`` (SHA-256 prehash).

    ``require_low_s`` (L6, parity with the TS ``verifyP256`` option) rejects the
    malleable high-S signature form. OFF by default: the Rust provider and Apple
    CryptoKit do NOT normalise to low-S today (~50% of real signatures are
    high-S), so enforcing it would fail genuine signatures. Enabling it by
    default is a follow-up gated on the producers normalising first."""
    pub_raw = base64.b64decode(public_key_b64)
    if len(pub_raw) != 64:
        return False
    try:
        sig = base64.b64decode(signature_der_b64)
        if require_low_s:
            _r, s = decode_dss_signature(sig)
            if s > _P256_HALF_N:
                return False
        pub = ec.EllipticCurvePublicKey.from_encoded_point(ec.SECP256R1(), b"\x04" + pub_raw)
        pub.verify(sig, message, ec.ECDSA(hashes.SHA256()))
        return True
    except (InvalidSignature, ValueError):
        return False


def sign_p256(private_key: ec.EllipticCurvePrivateKey, message: bytes) -> str:
    """Sign ``message`` (SHA-256 prehash) with a P-256 private key.

    Returns the DER-encoded signature, base64. Counterpart to ``verify_p256``
    — producers (the Python provider agent) use this; verifiers use
    ``verify_p256``.
    """
    signature = private_key.sign(message, ec.ECDSA(hashes.SHA256()))
    return base64.b64encode(signature).decode("ascii")


def verify_attestation_signature(attestation: Mapping[str, Any], public_key_b64: str) -> bool:
    """Verify an attestation's ``selfSignature`` against its own ``publicKey``.

    Authenticates every posture field (cdHash, getTaskAllow, encryptionPubKey,
    …). Strips ``selfSignature`` and ``$type``, canonicalizes the rest.
    """
    sig = attestation.get("selfSignature")
    if not sig:
        return False
    body = {k: v for k, v in attestation.items() if k not in ("selfSignature", "$type")}
    return verify_p256(public_key_b64, sig, canonical_bytes(body))


def verify_receipt_signature(receipt: Mapping[str, Any], attestation_public_key_b64: str) -> bool:
    """Verify a receipt's ``enclaveSignature`` against an attestation publicKey."""
    sig = receipt.get("enclaveSignature")
    if not sig:
        return False
    body = {k: v for k, v in receipt.items() if k not in ("enclaveSignature", "$type")}
    return verify_p256(attestation_public_key_b64, sig, canonical_bytes(body))
